File: src/db/central_db.py
import psycopg
from psycopg.rows import dict_row
from psycopg_pool import ConnectionPool
from contextlib import contextmanager
from src import config
from src import logger


# Create connection pool
connection_pool = ConnectionPool(
    conninfo=config.CENTRAL_DB_URL,
    min_size=1,
    max_size=25
)


def query(text, params=None):
    """Execute a query and return results as list of dicts."""
    with connection_pool.connection() as conn:
        conn.row_factory = dict_row
        with conn.cursor() as cursor:
            cursor.execute(text, params or [])
            if cursor.description:
                return cursor.fetchall()
            conn.commit()
            return []


def execute(text, params=None):
    """Execute a statement (INSERT, UPDATE, DELETE) and commit."""
    with connection_pool.connection() as conn:
        with conn.cursor() as cursor:
            cursor.execute(text, params or [])
            conn.commit()
            return cursor.rowcount


@contextmanager
def with_transaction():
    """Context manager for database transactions."""
    with connection_pool.connection() as conn:
        conn.row_factory = dict_row
        conn.autocommit = False
        cursor = conn.cursor()

        class TransactionClient:
            def __init__(self, cursor, conn):
                self.cursor = cursor
                self.conn = conn

            def query(self, text, params=None):
                self.cursor.execute(text, params or [])
                if self.cursor.description:
                    return self.cursor.fetchall()
                return []

        client = TransactionClient(cursor, conn)
        try:
            yield client
            conn.commit()
        except Exception as e:
            conn.rollback()
            raise e
        finally:
            cursor.close()


def shutdown_pool():
    """Close all connections in the pool."""
    try:
        connection_pool.close()
    except Exception as e:
        logger.error('Failed to close central DB pool', err=e)

File: src/db/centralDb.js
const { Pool } = require('pg');
const config = require('../config');
const logger = require('../logger');

const pool = new Pool({
  connectionString: config.centralDbUrl,
  max: 25,
  idleTimeoutMillis: 30_000
});

const withTransaction = async (callback) => {
  const client = await pool.connect();

  try {
    await client.query('BEGIN');
    const result = await callback(client);
    await client.query('COMMIT');
    return result;
  } catch (error) {
    await client.query('ROLLBACK');
    throw error;
  } finally {
    client.release();
  }
};

const shutdownPool = async () => {
  try {
    await pool.end();
  } catch (error) {
    logger.error({ err: error }, 'Failed to close central DB pool');
  }
};

module.exports = {
  pool,
  withTransaction,
  query: (text, params) => pool.query(text, params),
  shutdownPool
};
